components/causal_ts/src/tso.rs

```diff
@@ -62,55 +69,205 @@ impl TsoBatch {
                 Ordering::Relaxed,
                 Ordering::Relaxed,
             ) {
-                Ok(_) => return Some(TimeStamp::compose(self.physical, logical)),
+                Ok(_) => {
+                    return Some((
+                        TimeStamp::compose(self.physical, logical),
+                        logical + 1 == self.logical_end,
+                    ));
+                }
                 Err(x) => logical = x,
             }
         }
         None
     }
 
     // `last_ts` is the last timestamp of the new batch.
-    pub fn renew(&mut self, batch_size: u32, last_ts: TimeStamp) -> Result<()> {
-        let (physical, logical) = (last_ts.physical(), last_ts.logical() + 1);
-        let logical_start = logical.checked_sub(batch_size as u64).unwrap();
+    pub fn new(batch_size: u32, last_ts: TimeStamp) -> Self {
+        let (physical, logical_end) = (last_ts.physical(), last_ts.logical() + 1);
+        let logical_start = logical_end.checked_sub(batch_size as u64).unwrap();
+
+        Self {
+            size: batch_size,
+            physical,
+            logical_end,
+            logical_start: AtomicU64::new(logical_start),
+        }
+    }
+
+    /// Count of available TSO in the batch.
+    pub fn count(&self) -> u32 {
+        self.logical_end
+            .saturating_sub(self.logical_start.load(Ordering::Relaxed)) as u32
+    }
+
+    /// The original start timestamp in the batch.
+    pub fn original_start(&self) -> TimeStamp {
+        TimeStamp::compose(self.physical, self.logical_end - self.size as u64)
+    }
+
+    /// The excluded end timestamp after the last in batch.
+    pub fn excluded_end(&self) -> TimeStamp {
+        TimeStamp::compose(self.physical, self.logical_end)
+    }
+}
+
+/// `TsoBatchList` is a ordered list of `TsoBatch`. It aims to:
+/// 1. Cache more number of TSO to improve high availability. See issue #12794.
+///    `TsoBatch` can only cache at most 262144 TSO as logical clock is 18 bits.
+/// 2. Fully utilize cached TSO when some regions require latest TSO (e.g. in the scenario of leader transfer).
+///    Other regions without the requirement can still use older TSO cache.
+struct TsoBatchList {
+    inner: RwLock<TsoBatchListInner>,
+
+    /// Number of available TSO.
+    /// Using signed integer for avoiding a wrap around huge value as it's not precisely counted.
+    tso_count: AtomicI32,
```

Comment:
Why should we need this field, since we only use this field to update `tso_usage`?